Sofia sid/shooter

Panda8188 opened the pull request:
s h o o t e r . j a v a
s h o o t e r c a l . j a v a
s h o o t e r c o n s t a n t s . ja v a

itsmesp8343 committed:
Set Up Shooter

-Shooter, ShooterCal and ShooterConstants
-Shoot (PID)
-Stop
-Method for setting pivot angle
-setLatchAngle: when latching in endgame

itsmesp8343 committed:
UNFINISHED shooter

-attempting position control

voidSevenSevenSix committed:
some modifications to the shooter code, didn't have time to finish, profilfedpidcontroller stuff

itsmesp8343 committed:
Shooter Subsystem

-added position control
-still need ready to shoot method

Panda8188 committed:
Finished Shooter.java

- refactored `latch()` to `setLatchPosition()` with set degrees of 180
- `checkLatch()` removed and changed to `atDesiredPosition()` for more general checking
- Speed for `shoot()` is now constant

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,97 @@
+      pivotMotor.getAbsoluteEncoder(Type.kDutyCycle);
+  private double pivotDesiredPosition = ShooterCal.STARTING_POSITION_DEGREES;
+
+  /** Double distance, Double angle for pivotMotor */
+  private InterpolatingDoubleTreeMap pivotAngleMap;
```
Units in comment

Panda8188 committed:
removed unused imports

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,97 @@
+    pivotAngleMap.put(Constants.PLACEHOLDER_DOUBLE, Constants.PLACEHOLDER_DOUBLE);
+  pivotAngleMap.put(Constants.PLACEHOLDER_DOUBLE, Constants.PLACEHOLDER_DOUBLE);
+
+    controllerA = motorA.getPIDController();
+    controllerB = motorB.getPIDController();
```
Set these in declaration instead of constructor

Panda8188 committed:
commented out unit tests (due to known bug) so build compiles successfully

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,97 @@
+    controllerA = motorA.getPIDController();
+    controllerB = motorB.getPIDController();
+
+    controllerA.setP(ShooterCal.MOTOR_A_kP);
+    controllerA.setI(ShooterCal.MOTOR_A_kI);
+    controllerA.setD(ShooterCal.MOTOR_A_kD);
+    controllerA.setFF(ShooterCal.MOTOR_A_kFF);
+
+    controllerB.setP(ShooterCal.MOTOR_B_kP);
+    controllerB.setI(ShooterCal.MOTOR_B_kI);
+    controllerB.setD(ShooterCal.MOTOR_B_kD);
+    controllerB.setFF(ShooterCal.MOTOR_B_kFF);
```
Move these into a init sparks function that checks revliberrors. Should also set coast mode and current limits.

balghane replied:
And set velocity conversion factor between rpm and rps if you want to do rps

Panda8188 replied:
Switched to RPM instead of RPS (RPS was just a placeholder value because I didn't know what units to use

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  private void controlPosition(double distance){
+    pivotController.setGoal(pivotAngleMap.get(distance));
+    pivotDesiredPosition = pivotAngleMap.get(distance);
```
Maybe only call `pivotAngleMap.get(distance)` once and give it a variable name "target pivot angle". Or swap the order of the lines and setGoal to the desired position

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,97 @@
+    pivotController.setGoal(pivotAngleMap.get(distance));
+    pivotDesiredPosition = pivotAngleMap.get(distance);
+    double armDemandVoltsA = pivotController.calculate(pivotMotorAbsoluteEncoder.getPosition());
+    double armDemandVoltsB =
+        ShooterCal.PIVOT_MOTOR_FF.calculate(pivotController.getSetpoint().velocity);
```
Let's use acceleration as well, do what we did in the elevator PR with previous timestamp and velocity

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,97 @@
+
+/** Set arm to latching position (up = 180deg). Modified controlPosition. **/
+  public void setLatchPosition(){
+    pivotController.setGoal(180);
+    pivotDesiredPosition = 180;
```
180 isn't necessarily right. Move to a variable and make it a placeholder

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,97 @@
+    pivotController.setGoal(180);
+    pivotDesiredPosition = 180;
+    double armDemandVoltsA = pivotController.calculate(pivotMotorAbsoluteEncoder.getPosition());
+    double armDemandVoltsB =
+        ShooterCal.PIVOT_MOTOR_FF.calculate(pivotController.getSetpoint().velocity);
```
Same as before with acceleration.

balghane commented on src/main/java/frc/robot/subsystems/shooter/ShooterCal.java:
```diff
@@ -0,0 +1,34 @@
+import frc.robot.Constants;
+
+public class ShooterCal {
+  public static final double MOTOR_A_kP = Constants.PLACEHOLDER_DOUBLE;
+  public static final double MOTOR_A_kI = Constants.PLACEHOLDER_DOUBLE;
+  public static final double MOTOR_A_kD = Constants.PLACEHOLDER_DOUBLE;
+  public static final double MOTOR_A_kFF = Constants.PLACEHOLDER_DOUBLE;
```
Can you combine these declarations and add a comment. Same for below sets
```
  /** Add a comment here about the units. Input RPM or RPS (whatever you're using for target units), output [-1,1] */
  public static final double MOTOR_A_kP = Constants.PLACEHOLDER_DOUBLE,
  MOTOR_A_kI = Constants.PLACEHOLDER_DOUBLE,
  MOTOR_A_kD = Constants.PLACEHOLDER_DOUBLE,
  MOTOR_A_kFF = Constants.PLACEHOLDER_DOUBLE;
```

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,97 @@
+      new CANSparkMax(RobotMap.SHOOTER_MOTOR_B_CAN_ID, MotorType.kBrushless);
+  private final CANSparkMax pivotMotor =
+      new CANSparkMax(RobotMap.SHOOTER_PIVOT_MOTOR_CAN_ID, MotorType.kBrushless);
+  private final AbsoluteEncoder pivotMotorAbsoluteEncoder =
+      pivotMotor.getAbsoluteEncoder(Type.kDutyCycle);
```
`pivotMotorAbsoluteEncoder` will have an offset from zero, so anywhere we get the position we should offset. Maybe make a private method getPivotPosition that gets from the encoder and uses the offset

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,97 @@
+    double armDemandVoltsC = ShooterCal.ARBITRARY_PIVOT_FEED_FORWARD_VOLTS * getCosineArmAngle();
+    pivotMotor.setVoltage(armDemandVoltsA + armDemandVoltsB + armDemandVoltsC);
+  }
+  /** Check if robot is in latch position (180deg) */
```
This does not just check latch position so this comment isn't accurate

Panda8188 committed:
Changes made

- initialized controllers outside the constructor
- implemented initSparks() for error checks and initwithretry
- implemented getPivotPosition() to account for offset and used that whenever we call .getPosition() on absolute encoder
- Changed controlPosition() and setLatchPosition() to account for and use acceleration when determining amount of voltage needed to be used for pivot motor
- ShooterCal and ShooterConstants updated
- new Constants variable SPARK_INIT_RETRY_ATTEMPTS

Panda8188 committed:
removed potential merge conflicts

Panda8188 committed:
fixed merge conflicts now?

Panda8188 committed:
Merge branch 'main' into sofia-sid/shooter

balghane commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,10 +16,10 @@ public final class Constants {
   public static final int PLACEHOLDER_INT = 0;
   public static final float PLACEHOLDER_FLOAT = 0;
+
+  public static final int SPARK_INIT_RETRY_ATTEMPTS = PLACEHOLDER_INT;
```
We can just set this to 5

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,150 @@
+
+import java.util.Optional;
+
+import javax.swing.text.html.Option;
```
Remove, probably an accidental auto-import

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,150 @@
+      new CANSparkMax(RobotMap.SHOOTER_PIVOT_MOTOR_CAN_ID, MotorType.kBrushless);
+  private final AbsoluteEncoder pivotMotorAbsoluteEncoder =
+      pivotMotor.getAbsoluteEncoder(Type.kDutyCycle);
+  private double pivotDesiredPosition = ShooterCal.STARTING_POSITION_DEGREES;
```
Add units to variable name

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,150 @@
+  }
+
+  public void stop() {
+    controllerA.setReference(0, ControlType.kVelocity);
+    controllerB.setReference(0, ControlType.kVelocity);
```
Can we just do `setVoltage` on the motor to 0. The current code will actively try to slow the shooter down, which is not needed and is a waste of battery. setVoltage to 0 will let it spin down at its own rate

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,150 @@
+      pivotMotor.getAbsoluteEncoder(Type.kDutyCycle);
+  private double pivotDesiredPosition = ShooterCal.STARTING_POSITION_DEGREES;
+
+  private double prevVelocity = 0;
```
Units? Rpm right?

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,150 @@
+
+    double armDemandVoltsA = pivotController.calculate(getPivotPosition());
+    double armDemandVoltsB = pivotController.calculate(getPivotPosition());
+    if (!prevTimestamp.isEmpty()) {
+      armDemandVoltsB += ShooterCal.PIVOT_MOTOR_FF.calculate(prevVelocity, pivotController.getSetpoint().velocity, timestamp - prevTimestamp.get());
+    }
```
Maybe `else` we do the `calculate` with just the `final double nextVelocityRpm = pivotController.getSetpoint().velocity`?

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,150 @@
+  }
+
+
+/** Set arm to latching position (up = 180deg). Modified controlPosition. **/
+  public void setLatchPosition(){
+    pivotDesiredPosition = ShooterConstants.LATCH_ANGLE_DEGREES;
+    pivotController.setGoal(pivotDesiredPosition);   
+    double timestamp = Timer.getFPGATimestamp();
+
+    double armDemandVoltsA = pivotController.calculate(getPivotPosition());
+    double armDemandVoltsB = pivotController.calculate(getPivotPosition());
+    if (!prevTimestamp.isEmpty()) {
+      armDemandVoltsB += ShooterCal.PIVOT_MOTOR_FF.calculate(prevVelocity, pivotController.getSetpoint().velocity, timestamp - prevTimestamp.get());
+    }
+    double armDemandVoltsC = ShooterCal.ARBITRARY_PIVOT_FEED_FORWARD_VOLTS * getCosineArmAngle();
+
+    prevTimestamp = Optional.of(timestamp);
+    prevVelocity = pivotController.getSetpoint().velocity;
+    
+    pivotMotor.setVoltage(armDemandVoltsA + armDemandVoltsB + armDemandVoltsC);
+  }
```
This seems like a lot of repeated code with `controlPosition`. Can you think of a way to reduce the repeated code?

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,150 @@
+    errors += SparkMaxUtils.check(motorB.restoreFactoryDefaults());
+
+    errors += SparkMaxUtils.check(pivotMotor.setIdleMode(IdleMode.kBrake));
+    errors += SparkMaxUtils.check(motorA.setIdleMode(IdleMode.kBrake));
+    errors += SparkMaxUtils.check(motorB.setIdleMode(IdleMode.kBrake));
```
Let's set these to coast so they don't slow down too quickly

balghane commented on src/main/java/frc/robot/subsystems/shooter/ShooterCal.java:
```diff
@@ -0,0 +1,36 @@
+import frc.robot.Constants;
+
+public class ShooterCal {
+    /** Motor A & B & pivotMotor, Unit Inputs: RPM. output [-1,1] */
```
Can you add similar comments for the other sections? So that when people hover over `PIVOT_MOTOR_kI` for example, it'll show the comment

balghane commented on src/main/java/frc/robot/subsystems/shooter/ShooterCal.java:
```diff
@@ -0,0 +1,36 @@
+  public static final SimpleMotorFeedforward PIVOT_MOTOR_FF = new SimpleMotorFeedforward(
+            PIVOT_MOTOR_KS, PIVOT_MOTOR_KV, PIVOT_MOTOR_KA);
+
+  public static final double ARBITRARY_FEED_FORWARD_VOLTS = Constants.PLACEHOLDER_DOUBLE;
```
Not used, remove?

balghane commented on src/test/java/PoseBufferTest.java:
```diff
@@ -1,27 +1,27 @@
-import edu.wpi.first.math.geometry.Pose2d;
```
Can we un-remove the tests? It seems like sometimes it passes the unit tests on github and sometimes it does not. I saw the same issue on a PR I put up too

Panda8188 committed:
IT WORKS

- Uncommented tests beacuse they work
- MUST have  version 2024.2.1 wpilib for the tests to work
- minor bug fixes as per basel's requests
- 2nd iteration complete

balghane commented on src/main/java/frc/robot/subsystems/shooter/Shooter.java:
```diff
@@ -0,0 +1,137 @@
+    pivotController.setGoal(angle);
+    double timestamp = Timer.getFPGATimestamp();
+
+    double armDemandVoltsA = pivotController.calculate(getPivotPosition());
+    double armDemandVoltsB = pivotController.calculate(getPivotPosition());
```
What going on here, why are we doing this twice

Panda8188 committed:
removed duplicate variable (whoops)

balghane commented:
/format

github-actions[bot] committed:
Formatting fixes
